### ISSUE TITLE
Ligature request: %>% pipe operator in R

### ISSUE THREAD

**Issue by jtth** [tonsky/FiraCode#473]
We use the pipe operator `%>%` a lot in the [Tidyverse](http://www.tidyverse.org) set of packages in the R programming language. It's a way of stringing together many functions, converting them from prefix to infix. It would be really nice to have a pipe ligature.

<img width="839" alt="screen shot 2017-08-10 at 12 24 36 pm" src="https://user-images.githubusercontent.com/115188/29181018-9e97263e-7dc7-11e7-9b49-1059feadbb35.png">

**Comment by tonsky** [comment]
There was a couple of request for getting better coverage for R. I’m not against it, but can’t think of a way to improve the rendering of those

**Comment by jtth** [comment]
I'll give a design a shot in a day or so.

**Comment by msberends** [comment]
@jtth Want to share your design idea? Would be awesome to have this in R.

**Comment by bcollier** [comment]
I would second this, and third this for colleagues who have commented on how ugly the pipe symbol is. Has anyone made progress on this?

**Comment by msberends** [comment]
Maybe something like this? 
![image](https://user-images.githubusercontent.com/31037261/31330228-bedc90d8-acdd-11e7-8ea6-bfc2adb3a314.png)
It's a horizontal invert of the arrow that is shown for `<-`, with the zeroes of the percent mark.

Better question is: does R (or RStudio that is) provide ligatures for more than 2 characters?
When typing `===`, I get: ![image](https://user-images.githubusercontent.com/31037261/31330311-22ad525a-acde-11e7-90a7-872efebd8c88.png). I don't know if it will understand something in place for `%>%`.

**Comment by tonsky** [comment]
I don’t think R authors had anything in mind when they chose `%>%`. Probably it’s just a random sequence of characters. I also think it’s unfixable, visually I mean

**Comment by MadcapJake** [comment]
You could rotate the two `%` and connect them to form a line with a `>` through it. That way you would still have three-character width too.

**Comment by yeedle** [comment]
I would propose to just use F# pipe ligature for `%>%`. In R, custom operators are defined between percent signs, so anyone can define an operator like so `%whatever% <- function(){ ...}`. R's pipe operator was [actually](https://cran.r-project.org/web/packages/magrittr/vignettes/magrittr.html) inspired by F#'s pipe and is supposed to have similar behavior. However, since arbitrary-character operators have to be wrapped in `%...`%, the pipe ended up looking like `%>%`/

**Comment by yeedle** [comment]
There are several other popular R operators mentioned in https://github.com/tonsky/FiraCode/issues/39. Not sure what can be done with them though.

**Comment by bcollier** [comment]
I would second the proposal to use the F# ligature for the R pipe %>% . With respect to the #39 issue, it would be nice to have some of the other symbols changed, but I think most R developers would tell you they use the %>% in maybe 1/4 of their lines of code, it is everywhere. I think it should have the highest priority.

There was an earlier comment that RStudio does not support three-character ligatures. I am not sure about RStudio 1.0, but in RStudio 1.1 the double headed arrow <<- correctly changes to the ligature for that symbol with Fira Code.

**Comment by j-f1** [comment]
The `%` characters should be preserved — one of the rules of Fira Code AFAIK is that all of the characters stay. What about rendering `%>%` as `%|>%`?

**Comment by msberends** [comment]
> There was an earlier comment that RStudio does not support three-character ligatures. I am not sure about RStudio 1.0, but in RStudio 1.1 the double headed arrow <<- correctly changes to the ligature for that symbol with Fira Code.

True!

> I think most R developers would tell you they use the %>% in maybe 1/4 of their lines of code, it is everywhere

True!!!!

**Comment by bcollier** [comment]
j-f1, could you elaborate on rule about all of the characters staying?

**Comment by traversc** [comment]
Would like this as well.  How about removing the slashes and extra dots in the percent symbols, something like:

`𐩑<𐩑`

**Comment by tonsky** [comment]
I want the ligatures to resemble original characters combination

**Comment by j-f1** [comment]
@bcollier 

> j-f1, could you elaborate on rule about all of the characters staying?

See above ⬆️ 🙂

**Comment by timgoodman** [comment]
I agree with the suggestion of something like the F# pipe ligature (right-pointing triangle).  The percent signs (which are really just there because of R's requirements for custom binary operators) could be reduced to individual small circles.

Something like this:
∘▷∘
except I would stretch the triangle so it touches the circles.

**Comment by msberends** [comment]
Yeah, that’s beautiful! And I agree about the percentages :+1:

**Comment by jtth** [comment]
@timgoodman that's a lot better than what I had in mind. As long as the midline of the ∘ aligns with the centerline of the ▷ that's great!

Kind of like 
<img width="109" alt="screen shot 2017-12-10 at 12 17 29 pm" src="https://user-images.githubusercontent.com/115188/33807385-2212f9f2-dda4-11e7-80f9-26bcd63552f3.png">

**Comment by holstius** [comment]
1. Pilcrow
2. Manicule

**Comment by jtth** [comment]
@holstius You want a paragraph marker and a hand? What?

**Comment by holstius** [comment]
Either seemed appropriate to me.

%>% is read “and then”; it does the same thing pseudo-linguistically as a
paragraph break (that is, it separates “sentences”).

The “hand” is more of a visual analogy. There’s a rightward momentum.

On first read, I missed the point about preserving the elements of the
original characters. In that case, both my suggestions are way off the
mark.

**Comment by bwiernik** [comment]
One idea for all of these would be to drop the second o on the first % and the first o on the second %, then elongate the center operator character.

**Comment by msberends** [comment]
@tonsky Any good arguments against
![image](https://user-images.githubusercontent.com/31037261/37849282-5edf49ce-2ed7-11e8-9f06-703392dd7362.png)
??

Else I think most of us would say: please implement 😉

**Comment by epspi** [comment]
@tonsky and others:
Any interest in some variants inspired by the ligature for `$>`?
![screenshot 2](https://user-images.githubusercontent.com/8306014/38038074-2c6415c0-3278-11e8-9f79-cb4f12f671f9.jpg)



I thought that a compromise between *no* resemblance to the original characters (as in the `o |> o`) and a full representation of the two `%` signs might be to turn the `/` in the percent sign into an arrow. That sort of nests the arrow "within" a percent.
![pipe2](https://user-images.githubusercontent.com/8306014/38038986-3391147c-327a-11e8-8d86-89dac95a191f.jpg)
![pipe3](https://user-images.githubusercontent.com/8306014/38038991-35ba63ac-327a-11e8-9725-a228ecdff7e3.jpg)

Even the compound pipe could be represented with uneven `<>`

**Comment by jtth** [comment]
@epspi I like yours better than mine. Very nice! That's more clear and visually unpacks a lot better.

**Comment by epspi** [comment]
Thanks @jtth 
Adding better compound pipe
![file-1d](https://user-images.githubusercontent.com/8306014/38041242-40e9b0e8-327f-11e8-96a7-d543bb976398.jpg)

**Comment by epspi** [comment]
Alternative based on arrows instead of `> <`
![dd](https://user-images.githubusercontent.com/8306014/38045794-dbead3f4-328b-11e8-8266-c2227314b9fa.jpg)

**Comment by BStudent** [comment]
- 1. Consistency with other languages is a good principle, so following F# is a sound choice
- 2. Not to stray too far off-topic, but that would suggest that %in% follows the ancient lead of LaTeX `\in` which inserts the &#8712; (isin) symbol, consistent with math in general.
- 3. FWIW, I believe this represents some specialized type of op-amp in the EE world:
![image](https://user-images.githubusercontent.com/19792522/39404022-826d4412-4b3e-11e8-96a7-e20c51df4097.png)

**Comment by timgoodman** [comment]
While *some* resemblance to the original characters is good (and we should follow the rule that each character in the original corresponds to *some* part of the ligature), I think it's more important that the *meaning* is conveyed (in the semantic sense, not the "what characters does it stand for" sense).  For instance, `!=` produces a ligature which doesn't bear much resemblance to an exclamation point, but that's OK because the ligature clearly represents the *meaning* "not equal". (Put another way, the ligature looks like the symbol that probably *would* have been used, if only that symbol existed as a key on the standard keyboard.)<br /><br />
For this reason, I still think the percentages should be de-emphasized, since they don't convey much meaning.  They are just required for defining any custom binary operator in R.  If not for this requirement of the R language, the author of magrittr might well have chosen `|>`, given that the magrittr pipe was actually inspired by F#'s pipe, `|>` (see here: https://www.r-statistics.com/2014/08/simpler-r-coding-with-pipes-the-present-and-future-of-the-magrittr-package/ ).<br /><br />
So, the magrittr pipe is not a random character sequence as suggested above; it's just an approximation of F#'s pipe that meets the rules of R.  And F#'s pipe, it seems clear to me, is meant to have some resemblance to an arrow pointing in the direction of piping (and this is consistent with the fact that the reverse pipe resembles an arrow pointing the other way).<br /><br />

With all that said, I don't particularly care for the suggestion using diagonal arrows, because instead of using an arrow from x to f to represent "x piped into f", they use an arrow pointed off at an oblique angle, which seems less connected to the meaning.  Moreover, while they increase the resemblance to the percent sign (the semantically-irrelevant part), they actually *decrease* the resemblance to the greater-than symbol (the semantically-important part).  Even the suggestion before that, which is closer to resembling the greater-than sign, seems like it's favoring a resemblance to the percent sign a bit too much in my opinion.<br /><br />
So, I would still prefer the closed triangle with dots on either end, which:<br /><br />
1. Resembles the greater-than sign, and more importantly resembles the *meaning* of the symbol: an arrow in the direction of piping<br /><br />
2. Includes but deemphasizes the semantically-irrelevant percent signs<br /><br />
3. Resembles the equivalent F# operator, which actually inspired the magrittr pipeline operator

**Comment by BStudent** [comment]
1. While I favor consistency, @timgoodman makes good points, and I think most people would agree that the "op amp" just _looks better_.   I certainly  don't think it's a good idea to preserve the % signs that bracket the operators - they are escape characters and an accident of history.
2. I do think we want to be consistent with the conventions of mathematics.  And  %in% has a well-defined mathematical meaning.  However if we respect the constraint that the ligature must be the same number of characters I'd be OK with
`%in%`      maps to:        o&#8712;o
(note that I'm too lazy above to make real circles and just used lower-case 'o')
3. Where the real soul searching starts is in the **_undoing_** of unfortunate accidents of history.  Specifically,` %%`, which is actually called the _remainder_ and the floor of integer division is called the _modulus_.  As in:
A = modulus * divisor + remainder
And yes, probably Backus' fault, as I believe it first showed up in Fortran, but for some reason, computer languages use the term 'mod' to designate the remainder.
In the long run, one reason ligatures are a great thing is that they can be used to close this gap, so it's probably a good idea not to use them to widen it.

**Comment by kelleyvanevert** [comment]
I would very much love to see ligatures for these R operators, but I think it is imperative to keep (each) ligature's shape within close proximity of the original shape of the operator, in line with what the rest of the Fira Code ligatures do. (They all just slightly move around, enlarge and reduce, and extend/warp the sub-shapes. The most drastic changes appear in >>= and |>, and still here you can see the original shapes.)

Because the choice of the percentage sign is awkward and unfortunate (as @BStudent notes) and has no semantic relation to the operator (as opposed to say >= and !=), we cannot and should indeed refrain from trying to "semanticize" them. Moreover, because entire words can be used between then percentage signs, as in %between% and %in%, such an attempt would only lead to weird solutions.

I would propose just a minor de-emphasization or subtle change of presentation of the percentage signs, for example:

![screenshot from 2018-05-07 14-02-46](https://user-images.githubusercontent.com/43432/39701322-0834fac2-5201-11e8-841d-e52073e02e11.png)

![fira_r](https://user-images.githubusercontent.com/43432/39701905-39098152-5203-11e8-99ca-d8de1558ea1f.png)


This obviously makes the operator more legible, while retaining the visual characteristic of the original operator (beit awkwardly chosen or not).

**Comment by j-f1** [comment]
How about changing the `>` in `%>%` into the `|>` ligature?

**Comment by kelleyvanevert** [comment]
@j-f1 : that replacement would not generalize well to `%in%` and `%between%`.

I'm convinced that Fira Code (or any other symbols-ligaturing font) should stay well away from the programming language's motivation & stylistic reasons for choosing certain operators. We're not here to redefine the syntax of R, but just to make it slightly more pleasing (read: legible) to the eye. Replacing `!=` with `≠` is not about replacing an arbitrary "is not equal" syntax to `≠`, but about working with the idea that is already captured in `!=`, namely that it visually looks somewhat similar to `≠`, and then trying to preserve the original shape somewhat.

There is no logic to the percentage-operator-syntax of R, that is a given; but Fira Code shouldn't try to 'create' this logic out of thin air (and probably upset some R-coders along the way), but rather try to make the syntax R-coders already work this on a day-to-day basis a bit more legible / pleasing to the eye.

**Comment by bwiernik** [comment]
> There is no logic to the percentage-operator-syntax of R, that is a given; but Fira Code shouldn't try to 'create' this logic out of thin air (and probably upset some R-coders along the way), but rather try to make the syntax R-coders already work this on a day-to-day basis a bit more legible / pleasing to the eye.

Absolutely agree. Your proposal to simply reduce the size of the % signs to de-emphasize them is perfect.

**Comment by kelleyvanevert** [comment]
Other ways of de-emphasizing could be proposed of course. The following is a bit whacky (and probably too much), but just as an example of a different presentation that doesn't semanticize, but just re-arranges slightly to create a eye-catching but recognizable version of the original thing.

![fira_r_2](https://user-images.githubusercontent.com/43432/39711543-4caa86a8-5220-11e8-98f9-99e8e9a3f5f5.png)

**Comment by timgoodman** [comment]
I like the suggestion of reducing the size of the percentages -- I would suggest though that the circles in the percentages could become individual solid dots (so each percent symbol looks like dot slash dot).  To me that seems clearer and less visually noisy, compared to having tiny circles with just a wee bit of white space in the middle. <br />
<br />
I also like the suggestion of still turning the `>` into the ligature for `|>`, for the reasons I gave above.  But if folks prefer to leave it more closely resembling the original `>`, my suggestion would be to at least stretch it horizontally a bit to suggest the shape of an arrow a bit more, which I think is more suggestive of the symbol's meaning.  In fact, we could preserve the original width of the full three-character combination, just shrink the percentages and stretch the greater-than to make up the difference.

**Comment by kelleyvanevert** [comment]
Yes, @timgoodman, that's an excellent suggestion, it looks good because it's very clean :) And widening the `>` does seem like the Fira Code think to do, and it looks good:

![fira_r_3](https://user-images.githubusercontent.com/43432/39727267-53bcc6cc-5252-11e8-9cc7-7b45bc8a2643.png)

I do believe the whole point of Fira Code is to have the ligatures take the same space as the original characters, even though that often means adding more space around a shrunk figure, so yes, that is necessary.

**Comment by kelleyvanevert** [comment]
Nice. I do think I've shrunk the percentages a bit too much though, displaying the text so big is different than displaying it regular size, where legibility is more of an issue.

=> What do other people think?

**Comment by kelleyvanevert** [comment]
By the way, this would make existing code look something like this:

![fira_r_4](https://user-images.githubusercontent.com/43432/39728638-4d7f588c-5258-11e8-97e0-2079e322f1ba.png)

Where the original would be:

![fira_r_4o](https://user-images.githubusercontent.com/43432/39728691-8bde7464-5258-11e8-9470-ba280680d8ab.png)

... which makes me think it's not legible enough, and is a bit unclear still. Really, those percentage signs are ugly, man :| Crazy R people.

**Comment by bwiernik** [comment]
The @timgoodman suggestion (with the wider > character for the pipe) looks great! I think perhaps with the mini-% signs aligned with the bottom line, rather than centered would be less jarring regarding legibility. 

From a quick review of packages, it looks like the following operators are in at least somewhat common use. I don't know how flexible the ligature system is if it could detect a pair of % % separated by an arbitrary number of characters and ligature in all such cases.

**From `base`**
%*%
%o%
%in%
%x%
%%
%/%

**From `tidyverse` and others**
%>%
%+%
%@%
%|%
%||%
%T>%
%$%
%<>%
%within%
%m+%
%--%
%+replace%
%??%
%,%

**From `igraph`**
%u%
%--%
%->%
%<-%
%c%
%du%
%m%
%s%

**From `foreach`**
%:%
%do%
%dopar%

**From `stringi`**
%s<%
%s+%

**From `data.table`**
%chin%
%inrange%
%like%
%between%

**From `assertthat`**
%has_name%
%has_attr%
%has_args%

**Others**
%over%
%nin%

**Comment by msberends** [comment]
Although I really like these ideas of the last few days, do we know if this is technically possible for the creator of this font?

**Comment by kelleyvanevert** [comment]
I do think so. I'm not versed in typography and all the crazy opentype features, but it seems that the technique that Fira Code uses is generalizable, and it already includes a 4-character ligature (####). As far as I can tell, it has to do with not just using the normal `liga` substitution table, but some kind of quircky method that can generalize.

See [gen_calt.clj](https://github.com/tonsky/FiraCode/blob/master/gen_calt.clj).

**Comment by timgoodman** [comment]
@robertmitchellv Agreed, the fact that the same approach works for any custom binary operator is definitely a plus.

But now I'm really wanting to think of a good use for the "flip the table over" operator. :)

**Comment by jzadra** [comment]
This preserves the % and the directionality, but also the concept of "pipe" in that it has an opening for the data to flow through.

![image](https://user-images.githubusercontent.com/25590353/41679517-b5e5dcfc-748b-11e8-92d5-ee42887033d0.png)

**Comment by bwiernik** [comment]
@jzadra That doesn’t really look like the % at all to me, and there is no indication that a second % is there. Also isn’t extensible to any other % operator.

**Comment by bwiernik** [comment]
If so, sarcasm is hard to detect through text.

**Comment by hadley** [comment]
In R, `%%` is used to create custom infix operators, and technically any sequence of characters can come in between the two `%` (even whitespace, although that's obviously a bad idea) . Given that, I'd recommend either:

* Provide ligatures for a small set of the most popular infix functions (i.e. `%*%`, `%in%`, `%%`, `%/%`, `%>%`)

* Provide a general ligature (if possible) for any sequence of alphabetic characters that come in between `%` and `%`. However, even if OpenType supports this, it may have too high a risk of interfering with syntax from other languages.

Personally, of the options presented so far, I like @kelleyvanevert's suggestion to use smaller percent signs the best. I think using `°▷°` is too risky as it looks very different to `%>%` and it's is likely to confuse people.

**Comment by msberends** [comment]
Wow, the great Hadley jumped in himself 😄 Thanks!

**Comment by BStudent** [comment]
@hadley, this touches on an issue concerning (what I'll dub) "mathematician" and "designer" viewpoints to reconcile, some things seem obvious while others seem hard (and adherents don't always agree which is which). The examples I'd offer as a math camper are 

1. %in% - look no further than LaTeX (actually the semantics speak for themselves to a math person)

2. %% - I righteously proclaim that wrongs can be righted by translating this to *remainder* or some variant, with attendant aesthetic challenges

3. %>% is a special challenge ... One can almost justify changing pipe semantics: really, this is function composition in reverse, to wit:
y <- x %>% h() %>% g() %>% f() 

Is in most cases effectively function composition:
f(g(h(x)))

In many other branches of math this is written with the circle operator:

f  o  g  o  h(x)

(Forgive me, I am doing this from a phone so I have to use lowercase o as a kind of  circle operator purchased on sale at Walmart)

Other than Maybe for  %in% I do not claim to have all the answers much less good ones, and there are perhaps counter-examples to my interpretation of the pipe as function composition.  
But I do think that with existing sins - such as referring to remainder as modulus with casual abandon - destroying the very fabric of society, we have some obligation to maintain a consistency between Computer Science and  the other branches of Mathematics, or at least not make the situation worse.

**Comment by BStudent** [comment]
@bwiernik @traversc Pretty sure it was intended to be extensible ...

**Comment by BStudent** [comment]
@bwiernik I agree with your statement of purpose ... and pipe is a different animal from %in% and %% ... what would be nice from both a ligature standpoint and to your concern is if all languages with a pipe-like operator could agree on a single ligature for the representation, independent of the horrifying assortment of ascii characters they have settled upon for the compiler.

**Comment by andrewcstewart** [comment]
Considering that `%...%` frames every type of custom operator, I wonder if they could be rendered as standard bookends to any such customer operator... something more akin to `[>]`, except as a complete box (or something) framing the actual operator symbol. The left and right `%` would always render into the same framing characters, and the internal operator symbols would include both a `_` floor and `‾` ceiling.  

So in the case of `%>%` this could render to something like `[ ▷ ]` (though within a full rectangle).
![image](https://user-images.githubusercontent.com/1420654/57115753-9c890400-6d05-11e9-8870-bec509ba885d.png)

![image](https://user-images.githubusercontent.com/1420654/57115778-d0642980-6d05-11e9-988d-dcd8905485a1.png)

![image](https://user-images.githubusercontent.com/1420654/57115909-862f7800-6d06-11e9-8894-d578e80e59e4.png)

![image](https://user-images.githubusercontent.com/1420654/57115940-cd1d6d80-6d06-11e9-8365-defc8af3d3b6.png)


Alternatively, the standard framing characters could resemble something like an ellipses `⋯▷⋯` 
![image](https://user-images.githubusercontent.com/1420654/57115632-f0471d80-6d04-11e9-9d50-8076d9dd2d20.png)

**Comment by BrashEQLibrium** [comment]
My two cents:

1. A ligature that won't confuse programmers who know the language will be good. It should feel natural. That is why the arrow ligature works so well.
2. Someone said that the centered de-emphasized percentage symbols are "jarring" and suggested percentage symbols be placed on bottom line. I disagree. This will make the symbol harder to recognize. The triad of shapes of different sizes and the symmetry of the centered percent symbols makes those ligatures pop out in the test that I saw.

**Comment by kklot** [comment]
as Fira already has ligatures for those symbols inside `%` such as `%->%` it seems to be a wasting of time trying to come up with new symbols every time. Following the deemphasise ideas, replacing the `%` sign with that of Courier New font look pretty readable for me, even for those cases without ligature like the modulo
<img width="428" alt="Screenshot 2019-06-13 at 13 54 06" src="https://user-images.githubusercontent.com/4459378/59434652-e6c0d300-8de3-11e9-8cda-44aaa22b1483.png">
I guess changing this to Fira Code is out of question? if so is there a way to customise this for personal usage?

**Comment by tonsky** [comment]
@kklot you can open ttf/otf files in any font editor and change them to your liking. To rebuild from sources there is https://github.com/tonsky/FiraCode/blob/master/script/build

**Comment by jzadra** [comment]
> @kklot you can open ttf/otf files in any font editor and change them to your liking. To rebuild from sources there is https://github.com/tonsky/FiraCode/blob/master/script/build

Is there any documentation about how to do this for contributors? I looked around the files in this repo but didn't see anything like that.

**Comment by tonsky** [comment]
@jzadra I’m afraid not. You’re on your own

**Comment by msberends** [comment]
> as Fira already has ligatures for those symbols inside % such as %->% it seems to be a wasting of time trying to come up with new symbols every time.

Ever since tidyverse, the `%>%` is really something else, that could use a lot more dedication than what you’re suggesting. I think that’s clear from this (way too long) thread and the ones of #41 and #195.

**Comment by msberends** [comment]
> I want the ligatures to resemble original characters combination (@tonsky)

Then maybe it’s time for an R specific fork...

**Comment by msberends** [comment]
The relevance of this discussion will perish, since we now know **base R** will come up with a pipe in their next version as a drop-in replacement for the `magrittr`/`dplyr` pipe. 

The pipe will probably be `|>`.

**Comment by bwiernik** [comment]
The pipe is only one of many inline operators in R. Anything between two % would be better with a ligature. 

(Last I read, the base R pipe was not planned to have the same behavior as the magrittr pipe unfortunately)

**Comment by msberends** [comment]
Whatever you like. This is the actual implementation and it looks like perfect drop-in to me: https://github.com/wch/r-source/commit/a1425adea54bcc98eef86081522b5dbb3e149cdc

(off-topic mode off)

**Comment by timgoodman** [comment]
@bwiernik    It looks like instead of writing `y %>% f(x, .)`, you'll be able to write `y |> \(.) f(x, .)` (or choose a different argument name instead of `.`)

**Comment by kklot** [comment]
the new pipe look so nice aligned
![Screenshot 2021-01-28 at 12 14 31](https://user-images.githubusercontent.com/4459378/106131072-7dc9d280-6162-11eb-8631-a0581cbc4016.png)

**Comment by matthewstrasiotto** [comment]
The new pipe doesn't replace magrittr's pipe - pls don't forget about this

**Comment by matthewstrasiotto** [comment]
Hi - I made something like @kelleyvanevert 's [suggestion for "whacky" percents](https://github.com/tonsky/FiraCode/issues/473#issuecomment-387220130) in my own fork -

> Yes, @timgoodman, that's an excellent suggestion, I think it looks good because it's very clean :) And widening the `>` does seem like the Fira Code think to do, and it looks good:
> 
> ![fira_r_3](https://user-images.githubusercontent.com/43432/39727267-53bcc6cc-5252-11e8-9cc7-7b45bc8a2643.png)

My implementation:

![image](https://user-images.githubusercontent.com/39424834/119591998-fabc3e80-be1a-11eb-98bd-23e4049d4dff.png)

The actual glyphs, spacing & size certainly need work - These things look great blown up but get teensy when viewed at 12pts it turns out, and the spacing on the pipes needs to be increased.

Personally, @kelleyvanevert 's suggestion for "whacky" percents (i personally think of them as "bookend" percents) has been my favourite so far.

## Discussion of approach of bookending in general

I like the idea of "bookending" the enclosed operator in general, but whatever is done at least needs to suggest the underlying character.

In the vein of @andrewcstewart 's [comment](https://github.com/tonsky/FiraCode/issues/473#issuecomment-488885969):

> Considering that `%...%` frames every type of custom operator, I wonder if they could be rendered as standard bookends to any such customer operator... something more akin to `[>]`, except as a complete box (or something) framing the actual operator symbol. The left and right `%` would always render into the same framing characters, and the internal operator symbols would include both a `_` floor and `‾` ceiling.

<details><summary>List of proposed sketches by andrewcstewart</summary>
 
> So in the case of `%>%` this could render to something like `[ ▷ ]` (though within a full rectangle).
> ![image](https://user-images.githubusercontent.com/1420654/57115753-9c890400-6d05-11e9-8870-bec509ba885d.png)
> 
> ![image](https://user-images.githubusercontent.com/1420654/57115778-d0642980-6d05-11e9-988d-dcd8905485a1.png)
> 
> ![image](https://user-images.githubusercontent.com/1420654/57115909-862f7800-6d06-11e9-8894-d578e80e59e4.png)
> 
> ![image](https://user-images.githubusercontent.com/1420654/57115940-cd1d6d80-6d06-11e9-8365-defc8af3d3b6.png)
> 
> ![image](https://user-images.githubusercontent.com/1420654/57116051-88460680-6d07-11e9-9219-897fb0293ed0.png)
> 
> ![image](https://user-images.githubusercontent.com/1420654/57116068-a6136b80-6d07-11e9-948a-f4a4c50bc384.png)
> 
> ![image](https://user-images.githubusercontent.com/1420654/57116099-dc50eb00-6d07-11e9-8984-831ce612a2de.png)
> 
> ![image](https://user-images.githubusercontent.com/1420654/57116187-731da780-6d08-11e9-862b-df74f96a40c1.png)
> 
> ![image](https://user-images.githubusercontent.com/1420654/57116278-fb9c4800-6d08-11e9-8f4c-6881648554a4.png)
> 
> Alternatively, the standard framing characters could resemble something like an ellipses `⋯▷⋯`
> ![image](https://user-images.githubusercontent.com/1420654/57115632-f0471d80-6d04-11e9-9d50-8076d9dd2d20.png)

</details>
I like the principle of the suggestion, but none of the proposed ligatures do a lot to suggest the underlying characters.
For me, that's a problem,

## Proposed bookend that's evocative of % sign

I propose:

![image](https://user-images.githubusercontent.com/39424834/119595108-5e953600-be20-11eb-91e6-24a09f2c9b72.png)

To describe it:

- De-empathize the whole % operator with lighter stroke width 
- Downplay the functional significance of % by shifting the "dots" to the extremities & scaling them down
- Mirror the rightmost % operator to invoke the idea of "closure"
- Potential add a fine upper or lower "border" to further suggest the idea of "closure"

The above shows two alternatives, either a fully vertical stem on the `%` , or a diagonal stem on the `%` .

Personally, I think a diagonal stem tracks better as its more suggestive of the underlying character.

To extend this:

![image](https://user-images.githubusercontent.com/39424834/119596633-e8460300-be22-11eb-830c-5328d2704422.png)

**Comment by matthewstrasiotto** [comment]
![image](https://user-images.githubusercontent.com/39424834/119601036-aa99a800-be2b-11eb-9c92-b1ac67f6b8f9.png)

**Comment by RubenVerg** [comment]
> ![sample 1](https://user-images.githubusercontent.com/1420654/57115753-9c890400-6d05-11e9-8870-bec509ba885d.png)

This should be the ligature for `[▷]`, I think.

**Comment by Kalaschnik** [comment]
R finally got a native pipe: https://www.r-bloggers.com/2021/05/the-new-r-pipe/
Which is: `|>` and Fira Code renders this as:
![image](https://user-images.githubusercontent.com/9831678/142734291-d90727f7-3685-41ec-8e8b-e447ec97245e.png)

Due to that character sequence, I would propose that that magrittr sequence should look similar, so the user can identify "pipes", but is also able to tell if it’s the base R implementation the tidyverse one.

**Comment by adRn-s** [comment]
I'm really looking forward to these ligatures. Even if they would make my R coding look like an alien sci-fi movie, they're better than nothing and surely any other R programmer will agree with me that even the worst-looking of all of these (whichever that is) is better than not having the ligature at all... any lack of aesthetics is more on R side (having to implement custom operators with 3 characters, etc.).

Sorry for necrobumping the issue, @tonsky, but since your last comment seems to be years old... Can you update/ expand your considerations on this? Thanks!

**Comment by jzadra** [comment]
Yes, I too am still hoping for a pipe ligature!  This is my vote:

> This preserves the % and the directionality, but also the concept of "pipe" in that it has an opening for the data to flow through.
> 
> ![image](https://user-images.githubusercontent.com/25590353/41679517-b5e5dcfc-748b-11e8-92d5-ee42887033d0.png)

**Comment by msberends** [comment]
> Not a fan of this one... once you see the dick, you cannot unsee it...

Thanks for that 😓😂

**Comment by tonsky** [comment]
I guess I can make a stylistic set for smaller percentage. Rest of the suggestions change the glyph too much, and I want it to remain recognizable. Would that make everyone happy?

<img width="190" alt="Screenshot 2022-02-16 at 15 04 24" src="https://user-images.githubusercontent.com/285292/154280778-3e456aed-bb8d-4e09-98aa-d8e23d19dfb2.png">

**Comment by jzadra** [comment]
I like a connected glyph so that it looks like a single character vs 3 characters.  But I'll be happy with anything, thanks for your work on this!

**Comment by kalaschnik** [comment]
I know this does not represent three characters in one glyph, but having the native pipe in mind, with the idea that magrittr’s pipe should be visually similar to the native one; yet, distinct; maybe this:
<img width="48" alt="image" src="https://user-images.githubusercontent.com/9831678/154355142-b9936592-cd07-418d-800e-3cf081d49f60.png">

feel free to hate it, I am not sure what my feelings are about it anyway... but I got @jzadra point, that I would be nice to have something more ligature-like. Three disjunct characters making a glyph feels wrong somehow, yet better than nothing...

**Comment by msberends** [comment]
This has been a discussion for almost 5 years now, with numerous suggestions, ideas and opinions.

Can we please agree on the suggestion of the font designer @tonsky [of today](https://github.com/tonsky/FiraCode/issues/473#issuecomment-1041526123)? Would help us all to *actually* get somewhere 🙂

**Comment by jzadra** [comment]
> This has been a discussion for almost 5 years now, with numerous suggestions, ideas and opinions.

> Can we please agree on the suggestion of the font designer @tonsky https://github.com/tonsky/FiraCode/issues/473#issuecomment-1041526123? Would help us all to actually get somewhere 🙂

Agreed.  Like I said, I'll be thrilled to have anything, and I love Firacode so I support whatever @tonsky feels is best!

**Comment by kalaschnik** [comment]
Absolutely agree. Ignore my shit.

**Comment by ryandward** [comment]
I wonder if it would be possible to write some kind of theme, in the interim, just to hide the `%`s on valid syntactic markers and just bold/capitalize the intervening symbol or something.

I wonder how hard this could be for vs-code or r-studio?

```r
dataframe %>% value %in% array
```
could be rendered as

```python
dataframe > value in array

**Comment by strazto** [comment]
> I wonder if it would be possible to write some kind of theme, in the interim, just to hide the `%`s on valid syntactic markers and just bold/capitalize the intervening symbol or something.
> 
> I wonder how hard this could be for vs-code or r-studio?

Probably not that hard, but does hiding characters with syntactical significance *actually* improve your code readability?

Personally I don't think hiding the glyph altogether is a particularly good idea, however, my experience is that it wasn't that hard to implement a new ligature, I think I just used the trial version of the Glyphs (Mac only) app to make my variation of this ligature.

Just to reiterate, I've implemented a version of this ligature that you can download and install:
https://github.com/tonsky/FiraCode/issues/473#issuecomment-848437541 

I'm not saying it's the canonical version, or even good, but I think it's better than nothing, and certainly better than hiding characters that change your code's meaning.

**Comment by bwu62** [comment]
Commenting here to subscribe to updates, and also to suggest this as a possibility:

![image](https://github.com/user-attachments/assets/755f558c-dde8-45ee-b532-bb7f5d2ccacf)

The exact proportions should obviously be fine tuned more but I think this might be a good starting point. The ligature is both identifiable as %>% but also cohesive as a single unit.

Other iterations that could be considered if people think the ligature isn't clear enough:

1. The circles in the percent symbols can be slightly horizontally offset from each other if desired.
2. The bars in the percent symbols (i.e. just the first and last third of the long line) can be slightly sloped.
3. Both of the above?

**Comment by jzadra** [comment]
I'm still really wanting a ligature for the `%>%` pipe. Is this request dead? Or is there still a chance?  How can I help?
